packages/actor-init-query/lib/HttpServiceSparqlEndpoint.ts

```diff
@@ -1,73 +1,344 @@
-const quad = require('rdf-quad');
 
-// Force type on Cluster, because there are issues with the Node.js typings since v18
-const cluster: Cluster = clusterUntyped;
+const DF = new DataFactory();
 
 /**
  * An HTTP service that exposes a Comunica engine as a SPARQL endpoint.
  */
 export class HttpServiceSparqlEndpoint {
-  public static readonly MIME_PLAIN = 'text/plain';
-  public static readonly MIME_JSON = 'application/json';
-
-  public readonly engine: Promise<QueryEngineBase>;
-
-  public readonly context: any;
-  public readonly timeout: number;
-  public readonly port: number;
-  public readonly workers: number;
-
-  public readonly freshWorkerPerQuery: boolean;
-  public readonly contextOverride: boolean;
-
-  public lastQueryId = 0;
+  protected readonly port: number;
+  protected readonly timeout: number;
+  protected readonly workers: number;
+  protected readonly context: QueryStringContext;
+  protected readonly invalidateCacheBeforeQuery: boolean;
+  protected readonly freshWorkerPerQuery: boolean;
+  protected readonly allowContextOverride: boolean;
+  protected readonly endpointPath: string = '/sparql';
+  protected readonly engineFactory: QueryEngineFactoryBase<QueryEngineBase>;
+  protected readonly engineWrapper: (actorInitQuery: ActorInitQuery) => QueryEngineBase;
 
   public constructor(args: IHttpServiceSparqlEndpointArgs) {
     this.context = args.context || {};
     this.timeout = args.timeout ?? 60_000;
     this.port = args.port ?? 3_000;
     this.workers = args.workers ?? 1;
+    this.invalidateCacheBeforeQuery = Boolean(args.invalidateCacheBeforeQuery);
     this.freshWorkerPerQuery = Boolean(args.freshWorkerPerQuery);
-    this.contextOverride = Boolean(args.contextOverride);
-
-    this.engine = new QueryEngineFactoryBase(
+    this.allowContextOverride = Boolean(args.allowContextOverride);
+    this.engineWrapper = actorInitQuery => new QueryEngineBase(actorInitQuery);
+    this.engineFactory = new QueryEngineFactoryBase(
       args.moduleRootPath,
       args.defaultConfigPath,
-      actorInitQuery => new QueryEngineBase(actorInitQuery),
-    ).create(args);
+      this.engineWrapper,
+    );
+  }
+
+  /**
+   * Start the HTTP service.
+   * @param {Writable} stdout The output stream to log to.
+   * @param {Writable} stderr The error stream to log errors to.
+   */
+  public run(stdout: Writable, stderr: Writable): Promise<void> {
+    return cluster.isPrimary ? this.runPrimary(stdout, stderr) : this.runWorker(stdout, stderr);
+  }
+
+  public async handleRequest(
+    stdout: Writable,
+    stderr: Writable,
+    request: IncomingMessage,
+    response: ServerResponse,
+    engine: QueryEngineBase,
+    mediaTypeFormats: Record<string, string>,
+    mediaTypeWeights: Record<string, number>,
+  ): Promise<void> {
+    // Attempt to reconstruct the original full request URL with protocol and host
+    const requestProtocol = <string> request.headers['x-forwarded-proto'] ?? 'http';
```

Comment:
This `x-forwarded-proto` stuff seems new and unrelated to this PR.
Is it really needed?
If so, could this be a separate PR?

Reply:
It could probably be a separate PR, yes. The goal was to allow using Comunica behind a proxy (such as nginx, with TLS handled by the proxy) where the endpoint host or protocol does not match the URI that Comunica is listening at, in which case the SPARQL SD triples would need to be based on the URI that the client application is talking to, rather than what Comunica is listening at.

Reply:
Very useful indeed!